### PR TITLE
KAFKA-14701; Move `PartitionAssignor` to new `group-coordinator-api` module

### PR DESCRIPTION
**Pull request by dajac**
This patch moves the `PartitionAssignor` interface and all the related classes to a newly created `group-coordinator/api` module, following the pattern used by the storage and tools modules.

I'll update the KIP when this PR is merged.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.coordinator.group.assignor;
+package org.apache.kafka.coordinator.group.api.assignor;
 
 import org.apache.kafka.common.Uuid;
```
Comment (chia7712) [inline group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: I noticed that not all interfaces have `@InterfaceStability.Unstable`. Could you share the context to me?

Comment (chia7712) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: For another, `GroupSpec#MemberSubscriptionSpec` return the interface but `GroupSpec#memberAssignment` return a map struct. If it is public API now, maybe return `MemberAssignment` is more flexible if we want to enrich it in the future?

**Review by chia7712 — COMMENTED** [review]
@dajac nice patch! two small comments are left, and please take a look at build error

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java (hunk 1 above)
Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: > I noticed that not all interfaces have @InterfaceStability.Unstable. Could you share the context to me?

Fixed. Thanks.

> For another, GroupSpec#MemberSubscriptionSpec return the interface but GroupSpec#memberAssignment return a map struct. If it is public API now, maybe return MemberAssignment is more flexible if we want to enrich it in the future?

Good question. Reusing `MemberAssignment` is not ideal because we want to have the ability to pass internal objects. In this case, it could be backed by the `Assignment`. So, we could think of using an interface though.

On thing to consider is whether the `MemberAssignment` returned by the assignor would always be the same as the one provided to the assignor. The actual assignment (the map) is for sure. What do you think?

Comment (chia7712) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: >  So, we could think of using an interface though.

+1

> On thing to consider is whether the MemberAssignment returned by the assignor would always be the same as the one provided to the assignor. The actual assignment (the map) is for sure. What do you think?

not sure whether I have got the point. The member assignment in `GroupAssignment` returned from `PartitionAssignor#assign` is the new assignment. Hence, it should be different to assignment in `GroupSpec`.

**Comment by dajac** [comment]
I fixed the build issue.

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java (hunk 1 above)
Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: You suggested to use `MemberAssignment` in your previous comment. `MemberAssignment` is already used in `GroupAssignment` returned by `PartitionAssignor#assign`. I am not sure if you meant to reuse the same `MemberAssignment` in both places.

Comment (chia7712) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: > You suggested to use MemberAssignment in your previous comment. MemberAssignment is already used in GroupAssignment returned by PartitionAssignor#assign. I am not sure if you meant to reuse the same MemberAssignment in both places.

ummm, I guess I did not notice MemberAssignment is used by `GroupAssignment` before :_

Anyway, having a interface to replace `Map` struct is good way to me. Or we can keep current version if the tag "unstable" give the room to modify those public stuff :)

Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: I am playing with this. I will suggest something a bit later today.

Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:19]: @chia7712 I gave it a go. You can check the last commit to see how it looks like. I also moved some classes between internal packages while here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java [hunk 2]
```diff
@@ -88,4 +90,5 @@ public String toString() {
             ", memberAssignment=" + memberAssignment +
```
Comment (chia7712) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java:90]: `toString` still uses previous naming `MemberSubscriptionSpecImpl`

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java [hunk 3]
```diff
@@ -160,7 +166,7 @@ public GroupAssignment build() throws PartitionAssignorException {
      */
     private void maybeRevokePartitions() {
         for (String memberId : groupSpec.memberIds()) {
-            Map<Uuid, Set<Integer>> oldAssignment = groupSpec.memberAssignment(memberId);
+            Map<Uuid, Set<Integer>> oldAssignment = groupSpec.memberAssignment(memberId).partitions();
```
Comment (chia7712) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java:169]: As we have `MemberAssignment` interface now, could we manipulate `MemberAssignment` instead of raw map? Also, maybe we can remove `isImmutableMap` by checking the impl of `MemberAssignment`. For example:

```java
            MemberAssignment oldAssignment = groupSpec.memberAssignment(memberId);
            MemberAssignment newAssignment = null;

            // The assignor expects to receive the assignment as an immutable map. It leverages
            // this knowledge in order to avoid having to copy all assignments.
            if (! (oldAssignment instanceof Assignment)) {
                throw new IllegalStateException("The assignor expect an immutable map.");
            }
```

**Review by chia7712 — COMMENTED** [review]
@dajac thanks for updated PR

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java (hunk 3 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java:169]: We can probably do further improvements but this goes beyond the scope of this PR. We can do them as follow-ups. Does it work for you?

Regarding the usage of raw maps, the assignors work with them at the moment so we cannot change it easily. We are working hard on refactoring the assignors to perform better. Once we are done with this, we can consider it. As the interface will remain unstable until 4.0, we can still do it afterwards.

Comment (chia7712) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java:169]: that makes sense to me :smile:

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java [hunk 4]
```diff
@@ -39,7 +40,7 @@ public class MemberSubscriptionSpecImpl implements MemberSubscriptionSpec {
      * @param subscribedTopicIds    The set of subscribed topic Ids.
      * @param memberAssignment      The current member assignment.
      */
-    public MemberSubscriptionSpecImpl(
```
Comment (chia7712) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java:42]: `Constructs a new {@code MemberSubscriptionSpecImpl}.` -> `Constructs a new {@code MemberSubscriptionAndAssignmentImpl}.`

(I can't add comment on line#37)

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java [hunk 5]
```diff
@@ -45,18 +45,18 @@ public interface GroupSpec {
     /**
      * Gets the member subscription specification for a member.
      *
-     * @param memberId          The member Id.
+     * @param memberId The member Id.
      * @return The member's subscription metadata.
      * @throws IllegalArgumentException If the member Id isn't found.
      */
-    MemberSubscriptionSpec memberSubscription(String memberId);
```
Comment (rreddy-22) [inline group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:52]: nit: why did we rename it to MemberSubscription?

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The partition assignment for a consumer group member.
+ */
+@InterfaceStability.Unstable
+public interface MemberAssignment {
+    /**
+     * @return Target partitions keyed by topic Ids.
```
Comment (rreddy-22) [inline group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java]: nit: Since this is a generic interface should we say assigned partitions keyed by topic Id?

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The partition assignment for a consumer group member.
+ */
+@InterfaceStability.Unstable
+public interface MemberAssignment {
```
Comment (rreddy-22) [inline group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java:29]: Thanks for the change! I was wondering if this should be in a separate PR? Since it's not related to moving files to different modules and we're adding a new interface and making changes as a consequence?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberAssignmentImpl.java [hunk 8]
```diff
@@ -25,39 +26,39 @@
 /**
  * The partition assignment for a consumer group member.
  */
-public class MemberAssignment {
+public class MemberAssignmentImpl implements MemberAssignment {
     /**
      * The target partitions assigned to this member keyed by topicId.
      */
-    private final Map<Uuid, Set<Integer>> targetPartitions;
+    private final Map<Uuid, Set<Integer>> assignment;
```
Comment (rreddy-22) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberAssignmentImpl.java]: nit: should we name it partitions here as well? to keep it uniform?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/SubscribedTopicDescriberImpl.java [hunk 9]
```diff
@@ -29,14 +29,14 @@
  * The subscribed topic metadata class is used by the {@link PartitionAssignor} to obtain
  * topic and partition metadata for the topics that the consumer group is subscribed to.
  */
-public class SubscribedTopicMetadata implements SubscribedTopicDescriber {
+public class SubscribedTopicDescriberImpl implements SubscribedTopicDescriber {
```
Comment (rreddy-22) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/SubscribedTopicDescriberImpl.java:32]: nice! I like that everything is uniform with this naming

**Comment by rreddy-22** [comment]
Thanks for the PR!

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java [hunk 10]
```diff
@@ -45,18 +45,18 @@ public interface GroupSpec {
...
 
     /**
      * Gets the current assignment of the member.
      *
-     * @param memberId          The member Id.
-     * @return A map of topic Ids to sets of partition numbers.
-     *         An empty map is returned if the member Id isn't found.
+     * @param memberId The member Id.
+     * @return The member's assignment or an empty assignment if the
+     * member does not have one.
```
Comment (jeffkbkim) [inline group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java]: nit: i think keeping the indentation the same looks cleaner

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java [hunk 11]
```diff
@@ -29,7 +30,7 @@
 /**
  * An immutable assignment for a member.
  */
-public class Assignment {
+public class Assignment implements MemberAssignment {
```
Comment (jeffkbkim) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java:33]: could you help me understand the difference between Assignment and MemberAssignmentImpl?

Also, the name seems a bit vague given that it implements MemberAssignment

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java [hunk 12]
```diff
@@ -83,9 +85,10 @@ public int hashCode() {
 
     @Override
     public String toString() {
-        return "MemberSubscriptionSpecImpl(rackId=" + rackId.orElse("N/A") +
+        return "MemberSubscriptionAndAssignmentImpl(rackId=" + rackId.orElse("N/A") +
             ", subscribedTopicIds=" + subscribedTopicIds +
             ", memberAssignment=" + memberAssignment +
             ')';
     }
+
```
Comment (jeffkbkim) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/MemberSubscriptionAndAssignmentImpl.java]: nit: i think we can remove this

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java (hunk 5 above)
Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java:52]: Now that we have `MemberAssignment`, `MemberSubscriptionSpec` looked weird. So I renamed it to align them.

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java (hunk 7 above)
Comment (dajac) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java:29]: We could have done a separate PR but it is also fine doing it here as the main reviewer asked for it.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java (hunk 11 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java:33]: That's a good point. There are similar at the moment. Assignment should get extra fields in the future when we implement the client side assignor (if we ever do it) while MemberAssignment will stay as it is. We could consider merging them in the future.

**Comment by dajac** [comment]
@chia7712 @jeffkbkim @rreddy-22 Thanks for your comments. I have addressed them.

File: settings.gradle [hunk 13]
```diff
@@ -61,6 +61,7 @@ include 'clients',
     'examples',
     'generator',
     'group-coordinator',
+    'group-coordinator:group-coordinator-api',
```
Comment (mimaison) [inline settings.gradle:64]: We named other internal API modules just `api`, for example `connect:api`, `storage:api`. Should we stick to the same convention?

Comment (mimaison) [reply settings.gradle:64]: I just noticed we have `tools:tools-api` too :(

Comment (dajac) [reply settings.gradle:64]: Reusing `api` does not work. Gradle gets confused about it and dependencies fail. There is a bug in gradle. This is why we actually use `tools:tools-api` and why we have `project(":storage:api").name = "storage-api"` later in this file. So we effectively, use `module:module-api` as names expect for connect.

Comment (dajac) [reply settings.gradle:64]: For reference: https://github.com/gradle/gradle/issues/847

Comment (chia7712) [reply settings.gradle:64]: @mimaison you raise a good discussion. I had the same question in creating tools-api module 😄 

As new module, maybe storage:api can be rename to storage:storage-api for consistency.

File: group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java (hunk 7 above)
Comment (chia7712) [reply group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberAssignment.java:29]: oh, i neglected this PR aims to complete the migration, but I do love this change 😄

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java [hunk 14]
```diff
@@ -377,19 +375,19 @@ private Assignment newMemberAssignment(
...
-            return new Assignment(newMemberAssignment.targetPartitions());
+            return new Assignment(newMemberAssignment.partitions());
         } else {
             return Assignment.EMPTY;
         }
     }
 
     // private for testing
-    static MemberSubscriptionSpecImpl createMemberSubscriptionSpecImpl(
+    static MemberSubscriptionAndAssignmentImpl createMemberSubscriptionSpecImpl(
```
Comment (chia7712) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java]: Should it get renamed according to `MemberSubscriptionAndAssignmentImpl`? or it can be a constructor of `MemberSubscriptionSpecImpl`?

**Review by chia7712 — APPROVED** [review]
there is a little question, but this patch is good enough to me :)

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java (hunk 14 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java]: Fixed it!

**Review by chia7712 — APPROVED** [review]
LGTM!

**Review by rreddy-22 — APPROVED** [review]
Thanks! Looks good!

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java (hunk 11 above)
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/Assignment.java:33]: So, Assignment is different from MemberAssignmentImpl in that it will also be used for client side assignors, possibly in the future, whereas MemberAssignmentImpl is only used for server side assignors. is this correct?

I think it might be worth adding a short comment or perhaps merging it now in a follow up PR.

**Review by jeffkbkim — APPROVED** [review]

**Comment by dajac** [comment]
Merged to trunk and to 3.8.
